cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -1853,7 +1853,27 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             let src = input_to_reg_mem(ctx, inputs[0]);
             let dst = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
             let ty = ty.unwrap();
-            if ty.is_vector() {
+            if ty == types::I64X2 {
+                // This lowering could be a single instruction with AVX512F/VL's VPABSQ instruction.
+                // Instead, we use a separate register, `tmp`, to contain the results of `0 - src`
+                // and then blend in those results with `BLENDVPD` if the MSB of `tmp` was set to 1
+                // (i.e. if `tmp` was negative or, conversely, if `src` was originally positive).
+
+                // Emit all 0s into the `tmp` register.
+                let tmp = ctx.alloc_tmp(ty).only_reg().unwrap();
+                ctx.emit(Inst::xmm_rm_r(SseOpcode::Pxor, RegMem::from(tmp), tmp));
+                // Subtract the lanes from 0 and set up `dst`.
+                ctx.emit(Inst::xmm_rm_r(SseOpcode::Psubq, src.clone(), tmp));
+                ctx.emit(Inst::gen_move(dst, tmp.to_reg(), ty));
+                // Choose the subtracted lanes when `tmp` has an MSB of 1. BLENDVPD's semantics
+                // require the "choice" mask to be in XMM0.
+                ctx.emit(Inst::gen_move(
+                    Writable::from_reg(regs::xmm0()),
+                    tmp.to_reg(),
+                    ty,
+                ));
+                ctx.emit(Inst::xmm_rm_r(SseOpcode::Blendvpd, src, dst));
```

Comment:
@cfallin, I am suspicious about the `src.clone()` above as well as the `Writable::from_reg(regs::xmm0())`... am I breaking anything?

Reply:
I think this should be fine -- cloning the source is harmless (its live range will extend until its last use), and the explicit `Writable` coercion is fine with a fixed reg (we don't have a helper like `writable_xmm0()`). And no issues with the fixed reg use vs. src -- its live-range overlaps with src so the register allocator will know to move src elsewhere if originally in xmm0.